Add gtest macros and matchers for Result<T> and Error.

hsharma35 opened the pull request:
Summary: Adds matcher support for Result and Error types, similar to absl::Status.

Differential Revision: D77409312

larryliu0820 commented on runtime/core/testing_util/error_matchers.h:
```diff
@@ -0,0 +1,343 @@
+namespace internal {
+namespace {
+
+constexpr const char* toString(const Error error) {
```
Can you put this in Error? Would be easier for future changes to Error also change this API.

larryliu0820 commented on runtime/core/testing_util/test/error_matchers_test.cpp:
```diff
@@ -0,0 +1,134 @@
+/*
```
Naming convention is `test_error_matchers`I guess?

larryliu0820 commented on runtime/core/testing_util/test/targets.bzl:
```diff
@@ -0,0 +1,19 @@
+    TARGETS and BUCK files that call this function.
+    """
+    runtime.cxx_test(
+        name = "error_matchers_test",
```
```suggestion
        name = "test_error_matchers",
```

larryliu0820 reviewed (approved):
Can you add a test for real code (not `error_matchers_test.cpp`) that uses these matchers?

hsharma35 committed:
Add gtest macros and matchers for Result<T> and Error. (#12038)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/12038

Adds matcher support for Result and Error types, similar to absl::Status.

Reviewed By: larryliu0820

Differential Revision: D77409312
